### PR TITLE
Add OPA mediator for synapse extensions

### PR DESCRIPTION
**Pull request by 1akshitha**
This PR adds the OPA mediator. Git issue : https://github.com/wso2/product-apim/issues/11581

### CODE REVIEW

File: modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAUtils.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
+package org.apache.synapse.mediators.opa;
```
Comment (arunans23) [inline modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAUtils.java:1]: Shall we add the licence?

File: modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAMediatorFactory.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    static final QName TOKEN_Q = new QName(XMLConfigConstants.SYNAPSE_NAMESPACE, "accessToken");
+    static final QName POLICY_Q = new QName(XMLConfigConstants.SYNAPSE_NAMESPACE, "policy");
+    static final QName Rule_Q = new QName(XMLConfigConstants.SYNAPSE_NAMESPACE, "rule");
+    static final QName PAYLOAD_GENERATOR_Q = new QName(XMLConfigConstants.SYNAPSE_NAMESPACE, "requestGenerator");
+    static final QName ADVANCED_PROPERTIES_Q = new QName(XMLConfigConstants.SYNAPSE_NAMESPACE, "advancedProperties");
+    static final QName PROPERTY_Q = new QName(XMLConfigConstants.SYNAPSE_NAMESPACE, "property");
+    static final QName NAME_Q = new QName("name");
+
+
+
```
Comment (arunans23) [inline modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAMediatorFactory.java:28]: Shall we remove extra lines?

File: modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAUtils.java (hunk 1 above)
Comment (1akshitha) [reply modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAUtils.java:19]: done

File: modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAClient.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    private static final Log log = LogFactory.getLog(OPAClient.class);
+
+    private static final CloseableHttpClient httpClient = createHTTPClient();
+    public static final String AUTHORIZATION_HEADER = "Authorization";
+    public static final String BASIC = "Basic ";
+    public static final String CONTENT_TYPE_HEADER = "Content-Type";
+    public static final String APPLICATION_JSON = "application/json";
+
+    /**
+     * Method to publish the OPA payload to the OPA server
```
Comment (Nirothipan) [inline modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAClient.java:54]: method comments should end with a period.

File: modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAClient.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        }
+
+        HttpPost httpPost = new HttpPost(opaServerUrl);
+        httpPost.setHeader(CONTENT_TYPE_HEADER, APPLICATION_JSON);
+        if (credentials != null) {
+            httpPost.setHeader(AUTHORIZATION_HEADER, credentials);
+        }
+        try {
+            httpPost.setEntity(new StringEntity(payload));
+
```
Comment (Nirothipan) [inline modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAClient.java:77]: please remove these empty lines

File: modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAMediator.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    @Override
+    public boolean mediate(MessageContext messageContext) {
+
+        OPARequestGenerator requestGenerator = null;
+        String opaResponseString = null;
+        try {
+            try {
+                Class<?> requestGeneratorClassObject = Class.forName(requestGeneratorClassName);
```
Comment (Nirothipan) [inline modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAMediator.java:53]: don't we have the dependency for OPASynapseRequestGenerator here, why are we using dynamic loading?

File: modules/extensions/src/main/resources/META-INF/services/org.apache.synapse.config.xml.MediatorSerializer [hunk 6]
```diff
@@ -1,4 +1,5 @@
 org.apache.synapse.mediators.spring.SpringMediatorSerializer
 org.apache.synapse.mediators.bsf.ScriptMediatorSerializer
 org.apache.synapse.mediators.throttle.ThrottleMediatorSerializer
-org.apache.synapse.mediators.xquery.XQueryMediatorSerializer
\ No newline at end of file
+org.apache.synapse.mediators.xquery.XQueryMediatorSerializer
+org.apache.synapse.mediators.opa.OPAMediatorSerializer
```
Comment (Nirothipan) [inline modules/extensions/src/main/resources/META-INF/services/org.apache.synapse.config.xml.MediatorSerializer:5]: EOF

File: modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAMediator.java (hunk 5 above)
Comment (1akshitha) [reply modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAMediator.java:53]: We provide the option to Implement the same interface from APIM side. In that scenario, we provide the class name as a mediator input and the request generator object is created from that implementation.

File: modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAUtils.java (hunk 1 above)
Comment (1akshitha) [reply modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAUtils.java:19]: done

File: modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPASynapseRequestGenerator.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+        JSONObject opaPayload = new JSONObject();
+        Gson gson = new GsonBuilder().setPrettyPrinting().create();
+        String transportHeadersJson = gson.toJson(transportHeadersMap);
+
+        opaPayload.put("requestOrigin", requestOriginIP);
+        opaPayload.put("method", requestMethod);
+        opaPayload.put("path", requestPath);
+        opaPayload.put("httpVersion", requestHttpVersion);
+        opaPayload.put("transportHeaders", transportHeadersJson);
```
Comment (renuka-fernando) [inline modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPASynapseRequestGenerator.java:62]: above is the request payload sent to the OPA server, shall we use the same fields for Choreo Connect impl
@pubudu538  can u check

Comment (renuka-fernando) [inline modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPASynapseRequestGenerator.java:62]: Shouldn't we wrap this inside the field "inputs" in JSON?
eg:
```json
{
    "input": {
        "requestOrigin": "",
        "method": ""
    }
}
```

File: modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPASecurityException.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+public class OPASecurityException extends Exception {
+
+    public static final int MEDIATOR_ERROR = 90100;
+    public static final String MEDIATOR_ERROR_MESSAGE = "OPA Mediator: Unexpected mediator failure";
+    public static final int OPA_REQUEST_ERROR = 90101;
+    public static final String OPA_REQUEST_ERROR_MESSAGE = "OPA Mediator: Error with the OPA validation request";
+    public static final int OPA_RESPONSE_ERROR = 90102;
+    public static final String OPA_RESPONSE_ERROR_MESSAGE = "OPA Mediator: Error with the OPA validation response";
+    public static final int ACCESS_REVOKED = 90103;
+    public static final String ACCESS_REVOKED_MESSAGE = "OPA mediator: Access Revoked";
```
Comment (renuka-fernando) [inline modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPASecurityException.java:35]: Have we finalize these error codes, using the same codes for choreo connect

File: modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPARequestGenerator.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+     * Generate the OPA request payload from the provided message context and the additional Properties Map
+     *
+     * @param policyName         Name of the policy validated
+     * @param rule               The rule of the policy
+     * @param advancedProperties Advanced properties that can be used to construct the opa payload
+     * @param messageContext     The message to be validated with OPA server
+     * @return json input as a string and this will be sent to the OPA server for validation
+     * @throws OPASecurityException If an authentication failure or some other error occurs
+     */
+    String createRequest(String policyName, String rule, Map<String, Object> advancedProperties,
```
Comment (renuka-fernando) [inline modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPARequestGenerator.java:40]:
```suggestion
    String generateRequest(String policyName, String rule, Map<String, Object> advancedProperties,
```

File: modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPARequestGenerator.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+                         MessageContext messageContext) throws OPASecurityException;
+
+    /**
+     * Authenticates the given request using the authenticators which have been initialized.
+     *
+     * @param policyName     Name of the policy validated
+     * @param rule           The rule of the policy
+     * @param opaResponse    The message to be authenticated
+     * @param messageContext The message to be authenticated
+     * @return true if the authentication is successful (never returns false)
```
Comment (renuka-fernando) [inline modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPARequestGenerator.java:50]: return what OPA responds

File: modules/extensions/pom.xml [hunk 11]
```diff
@@ -327,6 +329,11 @@
             <classifier>runtime</classifier>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.json.wso2</groupId>
+            <artifactId>json</artifactId>
+            <version>${org.wso2.json.version}</version>
```
Comment (tharindu1st) [inline modules/extensions/pom.xml:335]: need to move to dependencymanagement

File: modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAUtils.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+            status = HttpStatus.SC_INTERNAL_SERVER_ERROR;
+            errorMessage = "Internal Sever Error";
+        }
+
+        messageContext.setProperty("HTTP_RESPONSE_STATUS_CODE", status);
+        messageContext.setProperty(SynapseConstants.ERROR_CODE, e.getErrorCode());
+        messageContext.setProperty(SynapseConstants.ERROR_MESSAGE, errorMessage);
+        messageContext.setProperty(SynapseConstants.ERROR_EXCEPTION, e);
+
+        Mediator sequence = messageContext.getSequence("_opa_policy_failure_handler_");
```
Comment (tharindu1st) [inline modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAUtils.java:89]: _opa_policy_failure_handler_ can we make it configurable

**Review by tharindu1st — CHANGES_REQUESTED** [review]

File: modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAUtils.java (hunk 12 above)
Comment (1akshitha) [reply modules/extensions/src/main/java/org/apache/synapse/mediators/opa/OPAUtils.java]: Fixed

File: modules/extensions/pom.xml (hunk 11 above)
Comment (1akshitha) [reply modules/extensions/pom.xml]: fixed via https://github.com/wso2/wso2-synapse/pull/1899/commits/d03967ca95a513fc1c5d958858bfdd887bcb3014

**Review by tharindu1st — APPROVED** [review]

**Pull request merged**